### PR TITLE
[REVIEW] Adding floating point specialization to comparators for NaNs

### LINKED ISSUE #3226: [FEA] Add floating point specialization to comparators for NaNs
**Is your feature request related to a problem? Please describe.**
Spark and cuDF Python expect `NaN` values to compare equally to one and another in sorting operations. The current libcudf++ comparators do not support this.

**Describe the solution you'd like**
A specialization needs to be added to `element_relational_comparator` for `is_floating_point<Element>` to check for `NaN` values using `isnan` and handle them appropriately.

This should also be done for `element_equality_comparator`.

### PR DESCRIPTION
**Pull request by rgsl888prabhu**
Added floating point specialization to comparators for NaNs and test cases to evaluate.

close #3226

### CODE REVIEW

File: cpp/include/cudf/table/row_operators.cuh [hunk 1]
```diff
@@ -71,8 +128,15 @@ class element_equality_comparator {
         return false;
       }
     }
-    return lhs.element<Element>(lhs_element_index) ==
-           rhs.element<Element>(rhs_element_index);
+    // NaNs are equal
```
Comment (harrism) [inline cpp/include/cudf/table/row_operators.cuh]: I don't think we should have to pay for this code for non-floating-point types. I think there should be a different operator for floating-point types of Element using SFINAE.

File: cpp/include/cudf/table/row_operators.cuh [hunk 2]
```diff
@@ -29,6 +29,63 @@
...
+*
+* @param[in] lhs first element
+* @param[in] rhs second element
+* @param[in] expected_ordering expected relation between elements
+* @returns bool always returns `false`
+*---------------------------------------------------------------------------**/
+template <typename Element,
+            std::enable_if_t<not std::is_floating_point<Element>::value>* = nullptr>
+__device__ bool evaluate_nan_ordering(Element lhs, Element rhs, weak_ordering expected_ordering) {
+    return false;
```
[the remark this answers is not in the document]
Comment (rgsl888prabhu) [reply cpp/include/cudf/table/row_operators.cuh]: It should be false as this function will be called by other types

File: cpp/include/cudf/table/row_operators.cuh [hunk 3]
```diff
@@ -29,6 +29,63 @@
...
+template <typename Element,
+            std::enable_if_t<std::is_floating_point<Element>::value>* = nullptr>
+__device__ bool evaluate_nan_ordering(Element const lhs, Element const rhs, weak_ordering expected_ordering) {
+
+    bool result = false;
+
+    switch(expected_ordering) {
+        case weak_ordering::EQUIVALENT: result =  std::isnan(lhs) and std::isnan(rhs);
+                                        break;
+        case weak_ordering::LESS: result = std::isnan(rhs) and not std::isnan(lhs);
```
[the remark this answers is not in the document]
Comment (rgsl888prabhu) [reply cpp/include/cudf/table/row_operators.cuh]: I am following this
`[-Inf, -ve, 0, -0, +ve, +Inf, NaN, NaN, null] (for null_order::AFTER)`
`[null, -Inf, -ve, 0, -0, +ve, +Inf, NaN, NaN] (for null_order::BEFORE)`

**Comment by jrhemstad** [comment]
> @jrhemstad So, we will have two sets of functions one for floating point and other for non-floating type.

Yes, but it's limited to the logic that needs to be specialized between floating/non-floating types.

File: cpp/include/cudf/table/row_operators.cuh [hunk 4]
```diff
@@ -29,6 +29,114 @@
...
+* @param[in] lhs first element
+* @param[in] rhs second element
+* @return weak_ordering Indicates the relationship between the elements in
+* the `lhs` and `rhs` columns.
+*---------------------------------------------------------------------------**/
+template <typename Element,
+            std::enable_if_t<std::is_floating_point<Element>::value>* = nullptr>
+__device__ weak_ordering relational_compare(Element lhs, Element rhs) {
+
+    if(std::isnan(lhs) and std::isnan(rhs)) {
```
Comment (jrhemstad) [inline cpp/include/cudf/table/row_operators.cuh]: While I don't expect there to be any difference... I think we should use the `isnan` from the CUDA math lib as opposed to from `std`. 

https://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH__SINGLE.html#group__CUDA__MATH__SINGLE_1gf8093cd7c372f91c9837a82fd368c711

**Comment by rgsl888prabhu** [comment]
@harrism

File: cpp/include/cudf/table/row_operators.cuh [hunk 5]
```diff
@@ -29,6 +29,114 @@
...
+        return weak_ordering::GREATER;
+    }
+    return weak_ordering::EQUIVALENT;
+}
+}
+/**---------------------------------------------------------------------------*
+* @brief A specialization for floating-point `Element` type rerlational comparison
+* to derive the order of the elements with respect to `lhs`. Specialization is to
+* handle `nan` in the order shown below.
+* `[-Inf, -ve, 0, -0, +ve, +Inf, NaN, NaN, null] (for null_order::AFTER)`
```
Comment (harrism) [inline cpp/include/cudf/table/row_operators.cuh]: Why does NaN appear twice in these?

Comment (jrhemstad) [reply cpp/include/cudf/table/row_operators.cuh]: To show `NaN == NaN`

**Pull request merged**